compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs

```diff
@@ -391,47 +600,68 @@ runEnv comp env0 = do
   return env1
 
 -- | Skolemise an expression variable and extend the environment.
-extVarEnv :: MonadEnv m ph
+extVarEnv :: (IsPhase ph, MonadEnv m ph)
   => ExprVarName
   -- ^ The expression variable to be skolemised.
   -> m ()
 extVarEnv x = extSkolEnv (SkolVar x)
 
 -- | Skolemise a list of record projection and extend the environment.
-extRecEnv :: MonadEnv m ph
+extRecEnv :: (IsPhase ph, MonadEnv m ph)
   => ExprVarName
   -- ^ The variable on which is being projected.
   -> [FieldName]
   -- ^ The fields which should be skolemised.
   -> m ()
-extRecEnv x fs = extSkolEnv (SkolRec x fs)
+extRecEnv x fs = do
+  env <- getEnv
+  let skols = envSkols env
+      curFs = [fs' | SkolRec x' fs' <- skols, x == x']
+      newFs = if null curFs
+        then fs
+        else fs ++ head curFs
```

Comment:
I don’t understand this. Why are you appending the fields of an existing skolem? And why only the first one?

Reply:
This is very confusing indeed. I added a note. It boils down to this:
- If you skolemise `x.[field1, field2]`, and skolemise `x.field3` afterwards, the end result should be that the list of skolems contains `x.[field1, field2, field3'.
- `extSkolEnv` works by just appending the new skolems to the front of the list, and ignoring any duplicate entries behind it, from now on. This means that adding `x.field3` would overwrite the previous entries.
- To fix this, we look up the latest entry for `x` in the skolems, and copy its fields into the new entry.